src/cache/algorithms/expr_variable_reordering.cpp

```diff
@@ -0,0 +1,253 @@
+// Rafael Sá Menezes - 04/2020
+
+#include "expr_variable_reordering.h"
+
+#include <cache/algorithms/algorithms_utils.h>
+
+namespace
+{
+constexpr std::array<expr2t::expr_ids, 4> binary_operations = {
+  expr2t::expr_ids::add_id,
+  expr2t::expr_ids::sub_id,
+  expr2t::expr_ids::mul_id,
+  expr2t::expr_ids::div_id};
+
+constexpr std::array<expr2t::expr_ids, 6> relations = {
+  expr2t::expr_ids::equality_id,
+  expr2t::expr_ids::notequal_id,
+  expr2t::expr_ids::lessthan_id,
+  expr2t::expr_ids::greaterthan_id,
+  expr2t::expr_ids::lessthanequal_id,
+  expr2t::expr_ids::greaterthanequal_id};
+} // namespace
+
+void expr_variable_reordering::run()
+{
+  auto expr_type = get_expr_type(expr);
+  switch(expr_type)
+  {
+  case PARSE_AS::BIN_OP:
+  {
+    this->run_on_binop(expr);
+    break;
+  }
+  case PARSE_AS::RELATION:
+  {
+    this->run_on_relation(expr);
+    break;
+  }
+  default:;
+  }
+}
+expr_variable_reordering::PARSE_AS
+expr_variable_reordering::get_expr_type(expr2tc &expr)
+{
+  static std::map<expr2t::expr_ids, PARSE_AS> expr_to_function;
+  static bool map_initialized = false;
+
+  if(!map_initialized)
+  {
+    for(auto i : binary_operations)
+    {
+      expr_to_function[i] = PARSE_AS::BIN_OP;
+    }
+
+    for(auto i : relations)
+    {
+      expr_to_function[i] = PARSE_AS::RELATION;
+    }
+
+    expr_to_function[expr2t::expr_ids::constant_int_id] = PARSE_AS::CONSTANT;
+
+    expr_to_function[expr2t::expr_ids::symbol_id] = PARSE_AS::SYMBOL;
+    map_initialized = true;
+  }
+
+  if(expr_to_function.find(expr->expr_id) == expr_to_function.end())
+  {
+    return PARSE_AS::SKIP;
+  }
+  return expr_to_function[expr->expr_id];
+}
+
+void expr_variable_reordering::run_on_binop(expr2tc &expr) noexcept
```

Comment:
Add a short description for every single method.

Reply:
The header includes the description of the method